winpr/libwinpr/path/path.c

```diff
@@ -1078,3 +1079,89 @@ const char* GetKnownPathIdString(int id)
 			return "KNOWN_PATH_UNKNOWN_ID";
 	}
 }
+
+static WCHAR* concat(const WCHAR* path, size_t pathlen, const WCHAR* name, size_t namelen)
+{
+	WCHAR* str = calloc(pathlen + namelen + 1, sizeof(WCHAR));
+	if (!str)
+		return NULL;
+
+	memcpy(str, path, pathlen * sizeof(WCHAR));
+	memcpy(&str[pathlen], name, namelen * sizeof(WCHAR));
+	return str;
+}
+
+BOOL winpr_RemoveDirectory_RecursiveA(LPCSTR lpPathName)
+{
+	WCHAR* name = ConvertUtf8ToWCharAlloc(lpPathName, NULL);
+	if (!name)
+		return FALSE;
+	const BOOL rc = winpr_RemoveDirectory_RecursiveW(name);
+	free(name);
+	return rc;
+}
+
+BOOL winpr_RemoveDirectory_RecursiveW(LPCWSTR lpPathName)
+{
+	BOOL ret = FALSE;
+
+	if (!lpPathName)
+		return FALSE;
+
+	const size_t pathnamelen = _wcslen(lpPathName);
+	const size_t path_slash_len = pathnamelen + 2;
+	WCHAR* path_slash = calloc(pathnamelen + 3, sizeof(WCHAR));
+	if (!path_slash)
+		return FALSE;
+	memcpy(path_slash, lpPathName, pathnamelen * sizeof(WCHAR));
+
+	const WCHAR sep[] = { PathGetSeparatorW(PATH_STYLE_NATIVE), '\0' };
+	const WCHAR star[] = { '*', '\0' };
+	PathCchAppendW(path_slash, path_slash_len, sep);
+	PathCchAppendW(path_slash, path_slash_len, star);
```

Comment:
Ah nvm. `path_slash_len` is actually `pathnamelen + 2;`. `path_slash_len` actually represents the destination buffer size, instead of the current `strlen for WChar` of the buffer.